File: leetcode-jvm/src/main/java/com/lone/leetcode/j/Solution7.java
package com.lone.leetcode.j;

/**
 * @author lone
 * @date 2018/9/30
 */
public class Solution7 {
    public int reverse(int x) {
        int ret = 0;
        while (x != 0) {
            int mod = x % 10;
            x /= 10;
            //正数溢出 mod > Integer.MAX_VALUE - Integer.MAX_VALUE / 10
            if (ret > Integer.MAX_VALUE / 10 || (ret == Integer.MAX_VALUE / 10 && mod > 7)) {
                return 0;
            }
            //负数溢出 mod < Integer.MIN_VALUE - Integer.MIN_VALUE / 10
            if (ret < Integer.MIN_VALUE / 10 || (ret == Integer.MIN_VALUE / 10 && mod < -8)) {
                return 0;
            }
            ret = ret * 10 + mod;
        }
        return ret;
    }


    public static void main(String[] args) {
        System.out.println(Integer.MAX_VALUE);
        System.out.println(Integer.MIN_VALUE);
        System.out.println(Integer.MAX_VALUE / 10);
        System.out.println(Integer.MIN_VALUE / 10);
    }
}

File: leetcode-jvm/src/main/kotlin/com/lone/leetcode/k/Solution7.kt
package com.lone.leetcode.k

import java.lang.Exception

/**
 * https://leetcode.com/problems/reverse-integer/
 * Given a 32-bit signed integer, reverse digits of an integer.
 * Example 1:
 *
 * Input: 123
 * Output: 321
 * Example 2:
 *
 * Input: -123
 * Output: -321
 * Example 3:
 *
 * Input: 120
 * Output: 21
 * Note:
 * Assume we are dealing with an environment which could only store integers within the 32-bit signed integer
 * range: [−231,  231 − 1]. For the purpose of this problem, assume that your function returns 0 when the reversed
 * integer overflows.
 * @author lone
 * @date  2018/9/30
 */
class Solution7 {
    fun reverse(x: Int): Int {
        if (x > -9 && x < 9) {
            return x
        }
        return try {
            if (x < 0) -1 * reverseInt(x.toString().substring(1)) else reverseInt(x.toString())
        } catch (e: Exception) {
            0
        }
    }

    private fun reverseInt(str: String): Int = str.reversed().toInt()
}
